### PR TITLE
tests: add spread tests for validation-sets

### LINKED ISSUE #4918: Test validation-set commands
### What needs to get done

Test `edit-validation-sets` and `list-validation-sets`

### Why it needs to get done

To test our code and prevent issues like #4909

### PR DESCRIPTION
**Pull request by mr-cal**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `tox run -m lint`?
- [x] Have you successfully run `tox run -e test-py310`? (supported versions: `py39`, `py310`, `py311`, `py312`)

-----

Adds a spread test for `edit-validation-sets` and `list-validation-sets`.

I need help from @sergiusens to set up a key for the `sergio.schvezov+snapcraft-test@canonical.com` account:
1. Create and register a key with the staging store. 
2. Add the private key as a Github secret called `SNAPCRAFT_ASSERTION_KEY`
    - The secret can be generated with `gpg --homedir ~/.snap/gnupg/ --armor --export-secret-keys testspreadkey | base64 | tr -d "\n"`

Fixes #4918
(CRAFT-3117)

### CODE REVIEW

File: tests/spread/general/validation-sets-new-key/task.yaml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  # XXX: register-key requires a manual login
+  snapcraft register-key default
+
+  # snapcraft will use a fake file editor
+  export EDITOR=editor.sh
+
+  # edit a validation set
+  snapcraft edit-validation-sets pv8nW1ZaULF7xXAkE3tiU3TdlOnYlGUr testvalidationset 1
+
+  snapcraft list-validation-sets
```
Comment (lengau) [inline tests/spread/general/validation-sets-new-key/task.yaml:39]: Can we check the output to ensure this set is included?

File: .github/workflows/spread.yml [hunk 2]
```diff
@@ -110,10 +110,13 @@ jobs:
...
         env:
           SPREAD_GOOGLE_KEY: ${{ secrets.SPREAD_GOOGLE_KEY }}
+          SNAPCRAFT_ASSERTION_KEY: "${{ secrets.SNAPCRAFT_ASSERTION_KEY }}"
           SNAPCRAFT_STORE_CREDENTIALS_STAGING: "${{ secrets.SNAPCRAFT_STORE_CREDENTIALS_STAGING }}"
           SNAPCRAFT_STORE_CREDENTIALS_STAGING_CANDID: "${{ secrets.SNAPCRAFT_STORE_CREDENTIALS_STAGING_CANDID }}"
           SNAPCRAFT_STORE_CREDENTIALS_STAGING_LEGACY: "${{ secrets.SNAPCRAFT_STORE_CREDENTIALS_STAGING_LEGACY }}"
-        run: spread google:ubuntu-22.04-64:tests/spread/general/store
+        run: |
+          spread google:ubuntu-22.04-64:tests/spread/general/store \
+            google:ubuntu-22.04-64:tests/spread/general/validation-sets
```
Comment (sergiusens) [inline .github/workflows/spread.yml:119]: can you move this outside of general and into a store suite?

File: tests/spread/general/validation-sets/task.yaml [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  rm -f store-key.txt
+
+execute: |
+  # ensure snapcraft is logged in and can access the store
+  snapcraft whoami
+
+  # snapcraft will use a fake file editor
+  export EDITOR="$PWD/editor.sh"
+
+  snapcraft edit-validation-sets pv8nW1ZaULF7xXAkE3tiU3TdlOnYlGUr testset 1 --key-name testspreadkey
```
Comment (sergiusens) [inline tests/spread/general/validation-sets/task.yaml:35]:
```suggestion
  snapcraft edit-validation-sets $(snapcraft whoami | yq .id) testset 1 --key-name testspreadkey
```

File: tests/spread/general/validation-sets/task.yaml [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  fi
+
+  # setup snap gpg dir
+  mkdir -p "$HOME/.snap/gnupg"
+  chmod 700 "$HOME/.snap/gnupg"
+
+  # import a registered key
+  echo "$SNAPCRAFT_ASSERTION_KEY" | base64 --decode > store-key.txt
+  gpg --homedir "$HOME/.snap/gnupg" --import store-key.txt
+  rm -f store-key.txt
```
Comment (sergiusens) [inline tests/spread/general/validation-sets/task.yaml:26]:
```suggestion
  rm -f store-key.txt
  
  snap install yq
```

File: tests/spread/general/validation-sets/task.yaml [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  STORE_API_URL: https://api.staging.snapcraft.io
+  STORE_UPLOAD_URL: https://storage.staging.snapcraftcontent.com
+  UBUNTU_ONE_SSO_URL: https://login.staging.ubuntu.com
+
+prepare: |
+  if [[ -z "$SNAPCRAFT_STORE_CREDENTIALS" ]]; then
+    echo "No credentials set in env SNAPCRAFT_STORE_CREDENTIALS"
+    exit 1
+  fi
+
```
Comment (sergiusens) [inline tests/spread/general/validation-sets/task.yaml:18]:
```suggestion

  if [[ -z "$SNAPCRAFT_ASSERTION_KEY" ]]; then
    echo "No gpg key set in env SNAPCRAFT_ASSERTION_KEY"
    exit 1
  fi
  
```

**Review by lengau — APPROVED** [review]
Nice!

File: tests/spread/store/validation-sets/task.yaml [hunk 6]
```diff
@@ -0,0 +1,43 @@
+summary: test the validation-set commands
+
+environment:
+  SNAPCRAFT_ASSERTION_KEY: "$(HOST: echo ${SNAPCRAFT_ASSERTION_KEY})"
+  SNAPCRAFT_STORE_CREDENTIALS: "$(HOST: echo ${SNAPCRAFT_STORE_CREDENTIALS_STAGING})"
+
+prepare: |
+  if [[ -z "$SNAPCRAFT_STORE_CREDENTIALS" ]]; then
+    echo "No credentials set in env SNAPCRAFT_STORE_CREDENTIALS"
+    exit 1
```
Comment (lengau) [inline tests/spread/store/validation-sets/task.yaml:10]:
```suggestion
    ERROR "No credentials set in env SNAPCRAFT_STORE_CREDENTIALS"
```

This doesn't need to include the trace.

File: tests/spread/store/validation-sets/task.yaml [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  fi
+
+  if [[ -z "$SNAPCRAFT_ASSERTION_KEY" ]]; then
+    echo "No gpg key set in env SNAPCRAFT_ASSERTION_KEY"
+    exit 1
```
Comment (lengau) [inline tests/spread/store/validation-sets/task.yaml:15]:
```suggestion
    ERROR "No gpg key set in env SNAPCRAFT_ASSERTION_KEY"
```

**Review by lengau — APPROVED** [review]

**Review by dariuszd21 — APPROVED** [review]

**Pull request merged**